### ISSUE TITLE
Failure to elimate intermediate heap allocations compared to clang

### ISSUE THREAD

**Issue by Kmeakin** [rust-lang/rust#106871]
This code allocates 3 `Box`s ([godbolt](https://godbolt.org/z/oEq1WarMr)):

```rust
pub fn foo(x: u32, y: u32) -> Box<u32> {
    let x = Box::new(x);
    let y = Box::new(y);
    Box::new(*x + *y)
}
```

The equivalent C code allocates just 1 `Box` ([godbolt](https://godbolt.org/z/nvddoqbWa)):
```c
uint32_t* foo(uint32_t x, uint32_t y) {
    uint32_t* p1 = malloc(sizeof(uint32_t));
    if (!p1) abort();
    *p1 = x;

    uint32_t* p2 = malloc(sizeof(uint32_t));
    if (!p2) abort();
    *p2 = y;

    uint32_t* p3 = malloc(sizeof(uint32_t));
    if (!p3) abort();
    *p3 = *p1 + *p2;
    
    return p3;
}
```

# Version
```
rustc 1.68.0-nightly (0b90256ad 2023-01-13)
binary: rustc
commit-hash: 0b90256ada21c6a81b4c18f2c7a23151ab5fc232
commit-date: 2023-01-13
host: x86_64-unknown-linux-gnu
release: 1.68.0-nightly
LLVM version: 15.0.6
```

**Comment by erikdesjardins** [comment]
Likely to be fixed by #102099

**Comment by clubby789** [comment]
After rebasing that PR onto current master, this issue is indeed fixed

@rustbot label +I-slow +A-llvm

**Comment by erikdesjardins** [comment]
This was fixed in 1.75 by https://github.com/rust-lang/rust/pull/102099: https://godbolt.org/z/f6b7W6n6W

**Comment by ChrisDenton** [comment]
Thanks! Closing as fixed then.
